### PR TITLE
[SPARK-41464][CONNECT][PYTHON] Implement `DataFrame.to`

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Implement `DataFrame.to` with a proto message

Implement `DataFrame.to` for scala API
Implement `DataFrame.to` for python API

### Why are the changes needed?
for Connect API coverage

### Does this PR introduce _any_ user-facing change?
'No'. New API

### How was this patch tested?
New test cases.

### CODE REVIEW

File: python/pyspark/sql/connect/dataframe.py [hunk 1]
```diff
@@ -1631,6 +1632,47 @@ def inputFiles(self) -> List[str]:
         query = self._plan.to_proto(self._session.client)
         return self._session.client._analyze(query).input_files
 
+    def to(self, schema: Union[DataType, str]) -> "DataFrame":
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/dataframe.py:1635]: PySpark API does not support this as a string. Let's get rid of string support for now.

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: python/pyspark/sql/connect/dataframe.py (hunk 1 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/dataframe.py:1635]: +1

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 2]
```diff
@@ -601,3 +602,19 @@ message Unpivot {
...
 }
+
+message ToSchema {
+  // (Required) The input relation.
+  Relation input = 1;
+
+  // (Required) The user provided schema.
+  //
+  // The Sever side will update the dataframe with this schema.
+  oneof schema {
```
Comment (zhengruifeng) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:613]: not related to this PR, I am wondering whether we can unify the schema with String Literal (DDL or JSON)

File: python/pyspark/sql/connect/dataframe.py (hunk 1 above)
Comment (beliefer) [reply python/pyspark/sql/connect/dataframe.py:1635]: OK

Comment (amaliujia) [reply python/pyspark/sql/connect/dataframe.py:1635]: yes we should have StructType in `pyspark_types_to_proto_types`.

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 3]
```diff
@@ -601,3 +602,19 @@ message Unpivot {
...
+
+    DataType datatype = 2;
```
Comment (amaliujia) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:615]: Maybe comment here to say this must be a StructType?

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto (hunk 2 above)
Comment (amaliujia) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:613]: This requires clients side to implement more. For example, client side always convert StructType to string representation then Spark will convert it back.

This basically asks clients to understand Spark's protocol on the string based schema and then implement it right?

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 4]
```diff
@@ -270,6 +271,24 @@ class SparkConnectPlanner(session: SparkSession) {
...
+  private def transformToSchema(rel: proto.ToSchema): LogicalPlan = {
+    val schemaType = if (rel.hasDatatype) {
+      DataTypeProtoConverter.toCatalystType(rel.getDatatype)
+    } else {
+      parseDatatypeString(rel.getDatatypeStr)
+    }
+
+    val schemaStruct = schemaType match {
+      case s: StructType => s
+      case d => StructType(Seq(StructField("value", d)))
```
Comment (amaliujia) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:283]: so it is not guaranteed that the client side always sends a StructType as schema when it is not a string?

File: python/pyspark/sql/connect/dataframe.py (hunk 1 above)
Comment (beliefer) [reply python/pyspark/sql/connect/dataframe.py:1635]: I have created https://github.com/apache/spark/pull/39103 let `pyspark_types_to_proto_types` support `StructType`.

**Comment by beliefer** [comment]
ping @zhengruifeng @amaliujia

File: python/pyspark/sql/tests/connect/test_connect_plan_only.py [hunk 5]
```diff
@@ -498,6 +499,19 @@ def test_coalesce_and_repartition(self):
             df.repartition(-1)._plan.to_proto(self.connect)
         self.assertTrue("numPartitions must be positive" in str(context.exception))
 
+    def test_to(self):
+        # SPARK-41464: test `to` API in Python client.
```
Comment (grundprinzip) [inline python/pyspark/sql/tests/connect/test_connect_plan_only.py:503]: please add a negative test case as well. Eg. trying to project a column that does not exist, uncastable data type etc.

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 6]
```diff
@@ -389,6 +389,21 @@ def test_schema(self):
             self.connect.sql(query).schema.__repr__(),
         )
 
+    def test_to(self):
+        # SPARK-41464: test DataFrame.to()
```
Comment (grundprinzip) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:393]: please add a negative test case as well. Eg. trying to project a column that does not exist, uncastable data type etc.

**Review by grundprinzip — APPROVED** [review]
Needs some additional test coverage on the e2e flow but otherwise good.

File: python/pyspark/sql/tests/connect/test_connect_basic.py (hunk 6 above)
Comment (amaliujia) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:393]: I was thinking if we only need rich tests in server side but on client side we just test a case pass through. 

However with another though I guess we need both sides have rich tests, even they may look redundant becaue
1. we need good coverage on client to_proto
2. we need good cover on server from_proto.

So maybe adding similar tests on both client and server sides.

Comment (grundprinzip) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:393]: The reason to have the tests on both sides is to understand the error propagation etc. For all assertions and checks we're adding on the client we need adequate tests for the branches.

Comment (beliefer) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:393]: OK

**Comment by zhengruifeng** [comment]
merged into master

**Comment by beliefer** [comment]
@zhengruifeng @grundprinzip @HyukjinKwon @amaliujia Thank you for all!

**Pull request closed without merging**